Add support for suggested BEP probe files

JuliaSprenger opened the pull request:
This PR adds the capability to export and import data from probeinterface to a probe description format as it is currently under discussion for a [preliminary BEP proposal](https://docs.google.com/document/d/1oG-C8T-dWPqfVzL2W8HO3elWK8NIh2cOCPssRGv23n0/edit#heading=h.7kenzwjry86m)

The proposed format for integration into BIDS is still under development, so changes are expected in the future. This PR is a prove of concept, that a translation between probeinterface and the proposed format is possible.

The code itself can probably simplified further in the future, but is accompanied already by some test.

JuliaSprenger commented:
The failing tests should be hopefully fixed by #36

samuelgarcia commented on probeinterface/io.py:
```diff
@@ -99,6 +99,292 @@ def write_probeinterface(file, probe_or_probegroup):
+                        'device_channel_indices']
+        df_others = df_probe.drop(used_columns, axis=1, errors='ignore')
+        for col_name in df_others.columns:
+            probe.annotate(**{col_name: df_probe[col_name].values})
```
Maybe we should also introduce an **array_annotations** as we do in neo for this, no ?

JuliaSprenger replied:
Yes, but this should be a separate PR.

samuelgarcia replied:
of course.

samuelgarcia commented on probeinterface/io.py:
```diff
@@ -99,6 +99,292 @@ def write_probeinterface(file, probe_or_probegroup):
+            raise ValueError(f'Can not generate {n} unique ids between {min} '
+                             f'and {max} in {trials} trials')
+        return ids
+
```
Instead of generating this in this function we could add a auto_generate_ids in ProbeGroup. 
Then here we could raise an error if the probegroup don't have electrode_ids with a message
"this probe group don't have ids, you should generate with ..."

JuliaSprenger replied:
Yes, this would be useful to add to the main functionality of probeinterface! For now I just generate unique integers, do yo have another preferred format for probe ids?

JuliaSprenger replied:
This functionality is now implemented on the probegroup level.

samuelgarcia commented:
This looks perfect.

samuelgarcia commented:
I merge in master the "electrode" to "contact" naming PR.
You should rebase this PR with it.

JuliaSprenger commented:
I rebased the branch, included the 'electrode' -> 'contact' name change and adjusted the BIDS reader and writer to use the probegroup `to_dataframe` and `from_dataframe` methods.

samuelgarcia commented:
Julia.
If it is finish I propose to merge this.

Then I would like to discuss about the generation of probes ids.

Then random way is OK for me in context of BIDS but for end user in some cases it could good to have something simple and readable like "probe0" and "probe0#contact10".

JuliaSprenger commented:
Yes, if you are happy with the changes, feel free to merge.

Regarding the generation of probe and contact ids we can extend the generation function if the topic pops up again.

samuelgarcia commented on probeinterface/probe.py:
```diff
@@ -620,7 +623,8 @@ def from_dataframe(df):
 
         if 'device_channel_indices' in df.columns:
-            probe.set_device_channel_indices(df['device_channel_indices'].values)
+            dev_channel_indices = df['device_channel_indices'].replace('', '-1')
```
dev_channel_indices is supposed to be integer index.

JuliaSprenger replied:
Ok, I didn't know. I added the corresponding restriction in `probe.set_device_channel_indices`, so it should be consistent independent of the origin of the indices now. Here in this line all input is still in string format, so I wouldn't mix string and int here.
